### PR TITLE
Add PoC Media Player with placeholder

### LINKED ISSUE #2422: Prototype unified AV player with placeholder
**Is your feature request related to a problem? Please describe.**
Initialising the player (SMP) for audio and video content has a large and immediate impact on page weight, e.g. `c0000000030o` has a total content size of 2.7MB (uncompressed) compared with 890KB when the player is disabled. This happens regardless of whether the user interacts with the player. We want to find a way of minimising the cost/performance impact on users without adversely affecting the user journeys for playing AV content. We think we can do this by delaying the embedding of the video till the user interacts with it.

We are also now moving to a unified approach for embedding AV content which uses the same approach for both AMP and canonical.

We need to prototype the front-end parts of this solution to ensure that it meets our UX, performance and a11y requirements.

**Describe the solution you'd like**
- Implement AV player component that supports both AMP and canonical
  - Use dummy embed URL for testing purposes as was done in https://github.com/bbc/simorgh/pull/1774
- Placeholder:
  - Canonical: implement a simple placeholder component for each piece of AV content using a static dummy image that 'upgrades' into a full player when the user clicks on it (see @rhell 's example at https://github.com/bbc/simorgh/pull/2550)
  - AMP: use the built-in placeholder behaviour provided by AMP custom elements
- (Canonical) evaluate ways of eliminating the need for the user to click multiple times (e.g. by using autoplay)
- The solution needs to be cross-browser compatible
- The solution needs to be fully accessible
    - Refer to screen reader UX in https://app.zeplin.io/project/5baa564c05c2eba0d2b89568/screen/5c596fa88971d341fae3fd53
- Summarise progress and main challenges found on the PR for review along with an evaluation of any options available

NB out of scope:
- UX finesse/formatting
- Unit tests
- Real data

**Describe alternatives you've considered**
We could stick with the existing solution or wait for such functionality to be available in the player itself.

**Testing notes**
- Must be a11y swarmed
- Must be cross-browser compatible

Dev insight: Will Cypress tests be required or are unit tests sufficient? Will there be any potential regression? etc

**Additional context**

- This should not be merged until A11Y swarm is done and tests have been carried out.

### PR DESCRIPTION
**Pull request by 12**
Resolves #2422 

**Prerequisites for reviewing**
- This is a **proof of concept**
- This component will be moved to Psammead in the very near future (it is in Simorgh so we can easily test it as a proof of concept in an article)
- Test coverage is already being addressed
- Early reviews on the approach and current implementation would help massively to ensure we have covered all basis
- There is a list of things below that this PR **will not do**, please keep that in mind when reviewing

**What this PR does do**
Creates a component that will embed an `iframe`, designed to display media content. 

For AMP, we use `amp-iframe` to display the video, and for canonical we show a placeholder (that matches the media holding image), that a user will click, which will then load an autoplay video from our media embed endpoint.

It also:
- Varies the embed URL based on environment
- Creates a container that is either 16:9 (landscape) or 9:16 (portrait) for the iframe to be embedded within
- Loads a placeholder, so we can defer the loading of the media iframe

**What this PR will not do**
- Work for portrait video
- Display subtitles (on any env that is not `live`, this is an issue with SMP)

**Code changes:**
- Moves `metadata` into `helpers` and create a corresponding `Metadata` component, to reduce complexity in core MediaPlayer component
- Rewrites the `MediaPlayerContainer`
- Calls the embed endpoint based on environment
- Removes nested grid and caption support (temporarily as there was a lot of complexity around nested grid that bulked up the MediaPlayer component)
- Rewrites snapshot tests

**NOTE: To get this working locally, you will have to run Simorgh via Mozart**

---

- [ ] I have assigned myself to this PR and the corresponding issues
- [ ] Tests added for new features
- [ ] Test engineer approval

### CODE REVIEW

File: src/app/containers/MediaPlayer/Amp.jsx [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    <React.Fragment>
+      <AmpHead />
+      <amp-iframe
+        sandbox="allow-scripts allow-same-origin"
+        layout="fill"
+        frameborder="0"
+        src={embedSrc}
+        allowfullscreen
+      >
+        <amp-img layout="fill" src={placeholderSrc} placeholder />
```
Comment (dr3) [inline src/app/containers/MediaPlayer/Amp.jsx]: Does this not need width & height to be valid amp?

**Comment by ryanmccombe** [comment]
Some input from media pod perspective (totally understanding this was made for articles first :D)

Amp, Canonical look great - we could use them pretty much as-is

Other components (most notably the media player container itself) seem to be highly dependant on getting an article block as their data.

To use this in our pod, we'd either have to:

- Let this just be the articles media player, and write our own container (we can still reuse the amp, canonical components though)
- Convert our data into a an articles block format (eww) before passing it in
- Heavily modify the media player container to make it generic

Totally understand that these tasks should probably fall to us though - shouldn't be articles' concern :D

File: src/app/containers/MediaPlayer/helpers/embedUrl.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
+const MORPH_SUFFIX = 'morph_env=test';
+const DOMAIN_PREFIX = {
+  local: 'http://localhost.bbc.com:8080/ws/av-embeds',
```
Comment (dr3) [inline src/app/containers/MediaPlayer/helpers/embedUrl.js]: Can this be in env config?

File: src/app/containers/MediaPlayer/helpers/embedUrl.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  test: 'https://www.test.bbc.co.uk/ws/av-embeds',
+  live: 'https://www.bbc.com/ws/av-embeds',
+};
+
+const embedUrl = (env, assetId, vpid, amp = false) => {
```
Comment (dr3) [inline src/app/containers/MediaPlayer/helpers/embedUrl.js]: This functions a bit hard to follow, can it be refactored at all to make it easier to follow?

File: src/app/containers/MediaPlayer/helpers/embedUrl.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
+const MORPH_SUFFIX = 'morph_env=test';
```
Comment (dr3) [inline src/app/containers/MediaPlayer/helpers/embedUrl.js]: Are we ok with this being in simorgh?

File: src/app/containers/MediaPlayer/index.jsx [hunk 5]
```diff
@@ -1,55 +1,49 @@
...
+  padding-top: ${({ orientation }) =>
+    orientation === 'Portrait' ? '177.78%' : '56.25%'};
+  position: relative;
+  overflow: hidden;
+`;
+
+const placeholderImage = src => {
+  const parts = src.split('/');
+  const [domain, media, imgService, width, ...extraParts] = parts;
+  const definedWidth = width.replace('$width', '512');
```
Comment (dr3) [inline src/app/containers/MediaPlayer/index.jsx:29]: srcset?

File: src/app/containers/MediaPlayer/index.jsx [hunk 6]
```diff
@@ -1,55 +1,49 @@
...
+const StyledContainer = styled.div`
+  padding-top: ${({ orientation }) =>
+    orientation === 'Portrait' ? '177.78%' : '56.25%'};
+  position: relative;
+  overflow: hidden;
+`;
+
+const placeholderImage = src => {
+  const parts = src.split('/');
+  const [domain, media, imgService, width, ...extraParts] = parts;
```
Comment (dr3) [inline src/app/containers/MediaPlayer/index.jsx:28]: Can we not reuse existing image url building here?

**Comment by 12** [comment]
@ryanmccombe Yeah it'd make sense to have a separate entry point for Articles and Media pages. We are planning on taking the actual component (without the Article specific logic) out of Simorgh and putting it into a new Psammead component.

**Review by dr3 — COMMENTED** [review]
Looks good :)

File: src/app/containers/MediaPlayer/helpers/embedUrl.js (hunk 4 above)
Comment (12) [reply src/app/containers/MediaPlayer/helpers/embedUrl.js]: This is being removed as we speak. It was to overcome the issue with mismatching our other applications, but that's now been resolved.

File: src/app/containers/MediaPlayer/index.jsx (hunk 5 above)
Comment (12) [reply src/app/containers/MediaPlayer/index.jsx:29]: Yeah this will be added at a later date 🙂

File: src/app/containers/MediaPlayer/index.jsx (hunk 6 above)
Comment (12) [reply src/app/containers/MediaPlayer/index.jsx:28]: It needs to be extended to support different origins, which imo should be a separate issue? Happy to create one if you agree 🙂

File: src/app/containers/MediaPlayer/Canonical.jsx [hunk 7]
```diff
@@ -0,0 +1,57 @@
+import React, { useState } from 'react';
+import { string } from 'prop-types';
+import styled from 'styled-components';
+import Image from '@bbc/psammead-image';
+
+const StyledPlaceholder = styled.div`
+  cursor: pointer;
+  position: absolute;
```
Comment (pjlee11) [inline src/app/containers/MediaPlayer/Canonical.jsx]: Could be wrong but the way the image placeholder works is that it has `position: relative;`, and the elements inside of the placeholder have `position: absolute;` - this differs from that approach

File: src/app/containers/MediaPlayer/index.jsx (hunk 6 above)
Comment (pjlee11) [reply src/app/containers/MediaPlayer/index.jsx:28]: I think the work to support different origins should be done as a precursor to this work

File: src/app/containers/MediaPlayer/Canonical.jsx (hunk 7 above)
Comment (pjlee11) [reply src/app/containers/MediaPlayer/Canonical.jsx]: `StyledContainer` is what I would define as the placeholder, so I'm curious what is the difference between `StyledContainer` and `StyledPlaceholder`

File: src/app/containers/MediaPlayer/index.jsx [hunk 8]
```diff
@@ -1,55 +1,49 @@
...
-    ChildWrapper = NestedGridItemChildSmall;
-    Container = GridItemConstrainedSmall;
-    wrapperSpan.default = '4';
-    portrait = true;
-  }
-
-  return { ParentWrapper, ChildWrapper, Container, wrapperSpan, portrait };
+const StyledContainer = styled.div`
+  padding-top: ${({ orientation }) =>
+    orientation === 'Portrait' ? '177.78%' : '56.25%'};
```
[the remark this answers is not in the document]
Comment (12) [reply src/app/containers/MediaPlayer/index.jsx]: It is gross, I agree. Again, something that can be refactored (somehow) in the future?

[the remark this answers is not in the document]
Comment (pjlee11) [reply src/app/containers/MediaPlayer/index.jsx]: Yeahhhhh, in the refactor either make them named constants or do the math for it or both using a comment. 

EG: `const landscapeRatio = '56.25%'; // (9/16)*100`

[the remark this answers is not in the document]
Comment (12) [reply src/app/containers/MediaPlayer/index.jsx]: I prefer the former to reduce unnecessary (albeit minute) computational power. That or just comment what it's doing

[the remark this answers is not in the document]
Comment (12) [reply src/app/containers/MediaPlayer/index.jsx]: Recent commit addresses this, implementing suggestion from @pjlee11

File: src/app/containers/MediaPlayer/Amp.jsx (hunk 1 above)
Comment (jamesdonoh) [inline src/app/containers/MediaPlayer/Amp.jsx]: You didn't look whether AMP can natively support the same 'click-to-embed' functionality we are offering for canonical did you? It would be a bit odd if our AMP pages ended up being heavier than canonical...

File: src/app/containers/MediaPlayer/Amp.jsx [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    </Helmet>
+  );
+};
+
+const Amp = ({ embedSrc, placeholderSrc }) => {
+  return (
+    <React.Fragment>
+      <AmpHead />
+      <amp-iframe
+        sandbox="allow-scripts allow-same-origin"
```
Comment (jamesdonoh) [inline src/app/containers/MediaPlayer/Amp.jsx]: Can anyone else get their head around this?

> If one does not opt into `allow-same-origin`, then every origin is allowed for the iframe. As soon as you add `allow-same-origin` to the sandbox the origin rules apply.

(from https://github.com/ampproject/amphtml/blob/master/spec/amp-iframe-origin-policy.md)

Reads to me that if you remove this value it should allow same-origin iframes, but that would defeat the whole point?

As a last resort we could set up a new DNS alias for `www` but that's quite unpleasant.

File: src/app/containers/MediaPlayer/Metadata.jsx [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+  return (
+    <Helmet>
+      {metadata && (
+        <script type="application/ld+json">{JSON.stringify(metadata)}</script>
+      )}
+    </Helmet>
+  );
+};
+
+export default Metadata;
```
Comment (jamesdonoh) [inline src/app/containers/MediaPlayer/Metadata.jsx:17]: Is this component even still needed?

Comment (12) [reply src/app/containers/MediaPlayer/Metadata.jsx:17]: As in, do we still need metadata added to the `head`, or do we need to separate it into its own component?

AFAIK we still need to add the structured data object to the head to improve SEO.

Comment (jamesdonoh) [reply src/app/containers/MediaPlayer/Metadata.jsx:17]: Oh yes, I was getting confused and thinking this was the `script` tag for the SMP settings object, but we already removed that right?

Comment (12) [reply src/app/containers/MediaPlayer/Metadata.jsx:17]: Yes that was a separate component called `AudioVideoHead`

File: src/app/containers/MediaPlayer/Amp.jsx (hunk 1 above)
Comment (12) [reply src/app/containers/MediaPlayer/Amp.jsx]: Nah 
> The fill layout is the only layout supported by amp-img that doesn't require any image dimensions

https://amp.dev/documentation/examples/style-layout/how_to_support_images_with_unknown_dimensions/

File: src/app/containers/MediaPlayer/Canonical.jsx (hunk 7 above)
Comment (12) [reply src/app/containers/MediaPlayer/Canonical.jsx]: The `StyledContainer` is the parent component that is sized to either 16:9 or 9:16. It needs a `position: relative` so that its child can be positioned `absolute`.

The `StyledPlaceholder` is the wrapper around the placeholder image. It adds a `cursor: pointer` to indicate it is a clickable element, and is sized to fill its parent container (the `StyledContainer`).

Comment (pjlee11) [reply src/app/containers/MediaPlayer/Canonical.jsx]: Spoke IRL. The container and placeholder seem to me that they could be a single element rather than 2 different ones. Also IMO the `StyledContainer` is acting as the placeholder to the video component, rather than a container - this naming could be reviewed

File: .storybook/webpack.config.js [hunk 11]
```diff
@@ -17,5 +17,12 @@ module.exports = {
         );
       },
     ),
+    new webpack.DefinePlugin({
+      'process.env': {
+        SIMORGH_EMBED_BASE_URL: JSON.stringify(
+          process.env.SIMORGH_EMBED_BASE_URL,
+        ),
+      },
+    }),
```
Comment (jamesdonoh) [inline .storybook/webpack.config.js:31]: Why is this needed/what is it doing?

File: src/app/containers/MediaPlayer/helpers/embedUrl.js [hunk 12]
```diff
@@ -0,0 +1,12 @@
+const embedUrl = (assetId, vpid, amp = false) => {
+  const baseUrl = process.env.SIMORGH_EMBED_BASE_URL;
```
Comment (jamesdonoh) [inline src/app/containers/MediaPlayer/helpers/embedUrl.js]: Let's start out with calling this `SIMORGH_AV_EMBED_BASE_URL` as before long we'll have other embeds (social etc.) that won't use the same endpoint.

File: src/app/containers/MediaPlayer/index.jsx [hunk 13]
```diff
@@ -1,55 +1,51 @@
...
-  NestedGridParentSmall,
-  NestedGridItemChildSmall,
-  NestedGridItemChildLarge,
-  GridItemConstrainedLargeNoMargin,
-  GridItemConstrainedSmall,
-} from '../../lib/styledGrid';
-
+import styled from 'styled-components';
+import Amp from './Amp';
+import Canonical from './Canonical';
```
Comment (jamesdonoh) [inline src/app/containers/MediaPlayer/index.jsx]: The naming convention in use elsewhere (well, Chartbeat and ATI) is `./amp` (for directory name) and `AmpChartbeatBeacon` (for component name), can we try to standardise? I don't mind what the standard is so long as it's fairly predictable. Happy for this to be done in a separate PR though.

File: src/app/containers/MediaPlayer/Amp.jsx (hunk 1 above)
Comment (jamesdonoh) [reply src/app/containers/MediaPlayer/Amp.jsx]: As discussed we can pick this up separately.

File: .storybook/webpack.config.js (hunk 11 above)
Comment (12) [reply .storybook/webpack.config.js:31]: Storybook does not have access to `process.env` variables, this adds the embed base URL to storybook.

**Review by jamesdonoh — APPROVED** [review]
This good stuff, we can iterate on the details but as this is disabled on live and we now have adequate test coverage around it I'd say it's OK to merge 👍

**Review by dr3 — APPROVED** [review]
Bit rough around the edges, but as a 'get it merged, its alpha' LGTM, great work
